guppylang/definition/pytket_circuits.py

```diff
@@ -0,0 +1,262 @@
+        try:
+            import pytket
+
+            if isinstance(self.input_circuit, pytket.circuit.Circuit):
+                try:
+                    import tket2  # type: ignore[import-untyped, import-not-found, unused-ignore]  # noqa: F401
+
+                    qubit = cast(TypeDef, globals["qubit"]).check_instantiate(
+                        [], globals
+                    )
+
+                    circuit_signature = FunctionType(
+                        [FuncInput(qubit, InputFlags.Inout)]
+                        * self.input_circuit.n_qubits,
+                        row_to_type([bool_type()] * self.input_circuit.n_bits),
+                    )
+
+                    if not (
+                        circuit_signature.inputs == stub_signature.inputs
+                        and circuit_signature.output == stub_signature.output
+                    ):
+                        raise GuppyError(PytketSignatureMismatch(func_ast, self.name))
+                except ImportError:
+                    err = Tket2NotInstalled(func_ast)
+                    err.add_sub_diagnostic(Tket2NotInstalled.InstallInstruction(None))
+                    raise GuppyError(err) from None
+            else:
+                raise GuppyError(PytketNotCircuit(func_ast))
+        except ImportError:
+            pass
+
+        return ParsedPytketDef(
+            self.id,
+            self.name,
+            func_ast,
+            stub_signature,
+            self.python_scope,
+            self.input_circuit,
+        )
+
+
+@dataclass(frozen=True)
+class ParsedPytketDef(CallableDef, CompilableDef):
+    """A circuit definition with parsed and checked signature.
+
+    Args:
+        id: The unique definition identifier.
+        name: The name of the function.
+        defined_at: The AST node where the function was defined.
+        ty: The type of the function.
+        python_scope: The Python scope where the function was defined.
+        input_circuit: The user-provided pytket circuit.
+    """
+
+    defined_at: ast.FunctionDef
+    ty: FunctionType
+    python_scope: PyScope
+    input_circuit: Any
+
+    description: str = field(default="pytket circuit", init=False)
+
+    def compile_outer(self, module: DefinitionBuilder[OpVar]) -> "CompiledPytketDef":
+        """Adds a Hugr `FuncDefn` node for this function to the Hugr."""
+        try:
+            import pytket
+
+            if isinstance(self.input_circuit, pytket.circuit.Circuit):
+                from tket2.circuit import (  # type: ignore[import-untyped, import-not-found, unused-ignore]
+                    Tk2Circuit,
+                )
+
+                circ = Hugr.load_json(Tk2Circuit(self.input_circuit).to_hugr_json())  # type: ignore[attr-defined, unused-ignore]
+                mapping = module.hugr.insert_hugr(circ)
+                hugr_func = mapping[circ.root]
+
+                # We need to remove input bits from both signature and input node.
```

Comment:
If I don't remove them, there is an runtime error while validating hugr along the lines of `The node Node(...) has an unconnected port Port(Incoming, 1) of type []+[].` where the node in question is `Call` (if I counted correctly) - I think this could maybe be fixed if we also explicitly add `bool`s to the function stub inputs, so it depends onwards whether we want that.